### ISSUE TITLE
Delete performance is orders of magnitude slower than batched inserts

### ISSUE THREAD

**Issue by eliaslevy** [rethinkdb/rethinkdb#4263]
It appears that deleting documents, even something as simple as deleting all documents in a table by just doing `table.delete.run`, is orders of magnitude slower than batch inserts.  This makes use of RethinkDB unfeasible in uses cases where there is a need to maintain a rolling set of documents in a table, such a time series, but where batch insert performance is the only mechanism to capable of inserting document at a sufficiently high rate.  Without being able to delete documents at the same rate as the batch inserts, the table would grown unbounded.

**Comment by mlucy** [comment]
Thanks for the report!  Sorry you ran into that.

To a certain degree this is expected because we picked the batch size for delete operations to stop them from interfering with other operations as much as possible, which will slow them down, but I talked to @timmaxw and he said you were seeing a factor of five difference, so there's probably something else going on as well.  We should look into this.

(@danielmewes, any insight on this?)

**Comment by danielmewes** [comment]
@eliaslevy a few questions to better understand what might be goign on:
* Is this on a cluster or a single server?
* If it's on a cluster, how many servers do you have, what's the network connection between them like, and how is the relevant table configured (number of replicas / number of shards)?
* Are there any secondary indexes defined on the table?
* What's the approximate size of the data? Does it fit into RAM or is it too large for that? (also are the disks SSDs or rotational disks?)
* Which version of RethinkDB are you running?

**Comment by eliaslevy** [comment]
@danielmewes Its a four server cluster.  Networking is dual bounded 5Gb/s PCIe ethernet interfaces.  I've run a number of tests with different number of tables (from 1 to 50) with 4 shards and 2 replicas. 5 secondary indexes.  Data fits on RAM easily (around 378GB RAM per box).  About 12GB of test input data with 10 million documents. Latest version of RethinkDB.

**Comment by danielmewes** [comment]
@eliaslevy Thanks for the information, we'll investigate...

**Comment by eliaslevy** [comment]
I should have also mentioned that the nodes were dual CPU systems with 10 cores per CPU and 2 threads per core, for 40 logical cores as observed by OS.

**Comment by danielmewes** [comment]
Actually assuming the table is sharded, this might be due to https://github.com/rethinkdb/rethinkdb/issues/4343 .
A batch insert would - assuming the new primary keys are not ordered or something - be processed across all four servers in parallel.
A `table.delete` query on the other hand would suffer from https://github.com/rethinkdb/rethinkdb/issues/4343 and would run deletions on only one server at a time.

**Comment by eliaslevy** [comment]
That may be a factor, but I suspect there is more to it.  The cluster had 4 nodes and IIRC the difference in wall clock was an order of magnitude longer rather than just 4X.

**Comment by danielmewes** [comment]
So it's slower by a factor of approximately 40x?

**Comment by eliaslevy** [comment]
No, I can't recall the exact numbers, but I think it was something like 10x or 8x.  I can try to regenerate them next week if necessary.

**Comment by danielmewes** [comment]
@eliaslevy Don't worry, we'll probably have to set up our own test anyway to track the issue down properly.
